pipeline/ingestor/ingestor_common/src/main/scala/uk/ac/wellcome/platform/ingestor/common/Indexer.scala

```diff
@@ -0,0 +1,82 @@
+package uk.ac.wellcome.platform.ingestor.common
+
+import com.sksamuel.elastic4s.ElasticDsl._
+import com.sksamuel.elastic4s.requests.bulk.{BulkResponse, BulkResponseItem}
+import com.sksamuel.elastic4s.requests.common.VersionType.ExternalGte
+import com.sksamuel.elastic4s.{ElasticClient, Index, Indexable, Response}
+import grizzled.slf4j.Logging
+import uk.ac.wellcome.elasticsearch.model.{CanonicalId, Version}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait Indexer[T] extends Logging {
+  val client: ElasticClient
+
+  implicit val ec: ExecutionContext
+  implicit val indexable: Indexable[T]
+  implicit val id: CanonicalId[T]
+  implicit val version: Version[T]
+  val indexName: Index
+
+  final def index(documents: Seq[T])
+    : Future[Either[Seq[T], Seq[T]]] = {
+
+    debug(s"Indexing ${documents.map(d => id.canonicalId(d)).mkString(", ")}")
+
+    val inserts = documents.map { document =>
+      indexInto(indexName.name)
+        .version(version.version(document))
+        .versionType(ExternalGte)
+        .id(id.canonicalId(document))
+        .doc(document)
+    }
+
+    client
+      .execute {
+        bulk(inserts)
+      }
+      .map { response: Response[BulkResponse] =>
+        if (response.isError) {
+          error(s"Error from Elasticsearch: $response")
+          Left(documents)
+        } else {
+          debug(s"Bulk response = $response")
+          val bulkResponse = response.result
+          val actualFailures = bulkResponse.failures.filterNot {
+            isVersionConflictException
```

Comment:
With this - is the idea that we might get a conflicting version from a source / VHS and if that's the case, we just don't care?

Could we log these to see how often it happens?

Reply:
This is the current behaviour: because SQS can send messages out of order the ingestor might receive version 2 before version 1. If that happens we don't want version 1 to override version 2. In that case, a version conflict exception is what you would get in elasticsearch. Since it's an expected scenario (although not frequent) we just log and move on. Logging is done [here]()https://github.com/wellcomecollection/catalogue/blob/709fdcf4d6967072455edfe1c619342928c3a2bf/pipeline/ingestor/ingestor_common/src/main/scala/uk/ac/wellcome/platform/ingestor/common/Indexer.scala#L74